projects/baby_driver/scripts/gpio_interrupts.py

```diff
@@ -0,0 +1,186 @@
+"""Python implementation of register_gpio_interrupt function"""
+
+from enum import IntEnum
+from collections import namedtuple
+
+import can_util
+from gpio_port import GpioPort
+import message_defs
+
+
+
+NUM_PINS_PER_PORT = 16
+
+GpioItInfo = namedtuple("GpioItInfo", ["port", "pin", "edge"])
+
+# pylint: disable=too-few-public-methods
+class InterruptEdge(IntEnum):
+    """Stores interrupt edge values"""
+    RISING = 0
+    FALLING = 1
+    RISING_AND_FALLING = 2
+    NUM_INTERRUPT_EDGES = 3
+
+# Callbacks that user defines when registering interrupt are stored here
+callback_dict = {
+    #("port", "pin"): callback,
+}
+
+def callback_listener(can_message):
+    """
+    Called by the notifier everytime a message is received from the firmware side notifiying
+    of an interrupt. It filters through the messages to find the can message with babydriver
+    message ID GPIO_IT_INTERRUPT and calls user-defined/default callback that is registered
+    with the interrupt
+    """
+
+    unfiltered_msg = can_util.Message.from_msg(can_message)
+
+    if (unfiltered_msg.message_id == message_defs.BABYDRIVER_CAN_MESSAGE_ID and
+        unfiltered_msg.data[0] == message_defs.BabydriverMessageId.GPIO_IT_INTERRUPT):
+
+        data = unfiltered_msg.data
+        port = data[1]
+        pin = data[2]
+        edge = data[3]
+
+        info = GpioItInfo(port, pin, edge)
+        # Calling function from callback_dict based on (port, pin)
+        try:
+            callback_dict[(port, pin)](info)
+        except TypeError as type_err:
+            raise TypeError ("Callback function parameters are of incorrect format (use a "
+                             "named tuple called info to store port, pin and edge)") from type_err
+
+
+def default_callback(info):
+    """ The function called after occurence of interrupt if
+       no user-defined callback function exists"""
+    info.port = chr(info.port + ord('A'))
```

Comment:
When this line executes, you'll get the error:
> AttributeError: can't set attribute

This is because you can't update the fields of a namedtuple like this; instead, just use a new variable.